File: leetcode/그래프/17/sol.java
// 풀이는 파이썬 코드에 있습니다.
// 파이썬으로 스케치된 코드를 자바로 포팅했습니다.
class Solution {
    ArrayList<String> answer = new ArrayList<String>();
    HashMap<Character, String> hm = new HashMap<>(); 
    
    public void dfs(int idx, String digits, String path) {
        if(path.length() == digits.length()) {
            answer.add(path);
            return;
        }
        
        int len = digits.length();
        for (int i = idx; i < len; ++i) {
            String temp = hm.get(digits.toCharArray()[i]);
            for (Character c : temp.toCharArray()) {
                dfs(i + 1, digits, path + c);
            }
        } 
    }
    
    public List<String> letterCombinations(String digits) {
        if(digits.length() == 0) {
            return answer;
        }
        
        hm.put('2', "abc");
        hm.put('3', "def");
        hm.put('4', "ghi");
        hm.put('5', "jkl");
        hm.put('6', "mno");
        hm.put('7', "pqrs");
        hm.put('8', "tuv");
        hm.put('9', "wxyz");
        
        dfs(0, digits, "");
        
        return answer;
    }
}

File: leetcode/그래프/17/sol.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        def dfs(idx, path) : 
            # 만약 각 output에 들어갈 path가
            # digits의 길이와 같으면 answer에 넣어준다.
            # 즉 끝까지 탐색이 완료된 경우면 answer에 넣어주고 백트래킹한다.
            if len(path) == len(digits) : 
                answer.append(path)
                return
            
            # 입력값 자릿수 단위로 이를 반복한다.
            for i in range(idx, len(digits)) :
                # 각 숫자에 해당하는 모든 문자열을 반복한다.
                for j in dic[digits[i]] : 
                    dfs(i + 1, path + j)
                    
                    
        # 예외처리
        if not digits : 
            return []
        
        dic = {"2" : "abc",
              "3" : "def",
              "4" : "ghi",
              "5" : "jkl",
              "6" : "mno",
              "7" : "pqrs",
              "8" : "tuv", 
              "9" : "wxyz"}
        
        answer = []
        dfs(0, "")
        
        return answer